### DISCUSSION TITLE
Backup and Restore MariaDB

### DISCUSSION THREAD

**Discussion by mcguirepr89** [mcguirepr89/BirdNET-Pi#140] [Show and tell]
Hello! This is a discussion dedicated to backing up and restoring the MariaDB. The steps I outline here are the recommended backup and restoration methods.

## Using the "Database Maintenance" Tool (Adminer)
You can very easily create a backup of your database using the "System" > "Tools" > "Database Maintenance" tool. 
### Create the backup
1. To do so, open the "Database Maintenance" tool and connect to the database (i.e., login to Adminer to connect to the database)
    ![image](https://user-images.githubusercontent.com/60325264/153716076-67bc099a-5193-4c93-9745-52e2c618bfb3.png)
1. Once connected, click the "Export" link on the left side of the window, which will bring you to the page you see below.
    ![image](https://user-images.githubusercontent.com/60325264/154132567-e2126af0-da09-4375-845e-897b96545ae7.png)
1. Choose "`gzip`" as the **Output**, "`SQL`" as the **Format**, and select "`DROP+CREATE`" for both the **Database** and the **Tables** leave everything else as is.
1. Finally, click **Export** and you will begin downloading a copy of your database.
### Restore the backup
1. To restore the backup, you will again use the "Database Maintenance" tool to connect to the database (login to Adminer)
    ![image](https://user-images.githubusercontent.com/60325264/153716219-27993497-c94f-47b3-95aa-5767abd7498e.png)
1. Once connected, click the "Import" link on the left side of the window, which will bring you to the page you see below
    ![image](https://user-images.githubusercontent.com/60325264/153716242-214af684-b15c-4e28-ba1d-06ae1a20bb36.png)
1. Click the "Choose Files" button to select the database backup you created (you will see the file name to the right of "Choose Files" after you have selected it).
    ![image](https://user-images.githubusercontent.com/60325264/153716281-1bb49555-7a62-467a-8e82-c5231727d6ed.png)
1. Finally, click **Execute** and you will see that the database has been restored.
    ![image](https://user-images.githubusercontent.com/60325264/153716334-a0c824b3-8dda-4f51-9980-bfdbc4236f93.png)
That's it! You restored the database!
## Using `mysqldump` and `mysql`
You can also create and restore backups in the command line, though it is not quite as easy as the method above.
### Create the backup
1. First, ensure you have your database password handy (preferably in the clipboard ready to paste).
1. With your database password handy, issue the command below, which will prompt you for the password:
    ```
    mysqldump -ubirder -p birds detections > /home/pi/BirdNET-Pi/scripts/detections.sql
    ```
    The command above created a backup of the `detections` table of the `birds` database and placed that file in the `scripts` directory.
1. Use the "System" > "Tools" > "File Manager" tool to navigate to the `scripts` directory and download a copy of the `detections.sql` file.
### Restore the backup
1. Using the "System" > "Tools" > "Database Maintenance" tool, upload the `detections.sql` backup file to the `scripts` directory.
1. In a terminal, change the permissions of the backup so that `pi` owns and can use the backup.
    ```
    sudo chown pi:pi /home/pi/BirdNET-Pi/scripts/detections.sql
    ```
1. Then restore the backup (the command below will prompt you for your _new_ installation's database password, so don't use the password from the old database).
    ```
    mysql -ubirder -p birds < /home/pi/BirdNET-Pi/scripts/detections.sql
    ```
That's it! You restored the database!

**Comment by Svardsten53** [comment]
Today I did a new installation and planned to restore the old database from a backup. The instructions above seems very simple and straightforward, but I failed. I tried both with Adminer tool and manual commands. I believe it's very important this instruction works. Better safe than sorry.


![Skärmbild 2022-02-15 190643](https://user-images.githubusercontent.com/62435779/154127117-515d66b3-c4ed-48e8-845a-dc77dc2a4bcf.png)
![Skärmbild 2022-02-15 193247](https://user-images.githubusercontent.com/62435779/154127176-2e41f5ee-be5b-45f9-b3b1-05d482192bb3.png)

**Reply by mcguirepr89** [reply]
On your new installation, login to the database (that is presumably empty) and delete the existing database.
Go to the "Alter Database" link near the top:
![image](https://user-images.githubusercontent.com/60325264/154131923-faaf5e01-b512-455f-ba7d-446a71a404d7.png)
![image](https://user-images.githubusercontent.com/60325264/154131948-9c3b40e7-4688-4e81-9bc0-407302cbe860.png)
And choose to "drop" the existing `birds` database.

Then try to import the backup and let me know if that worked. (There is a step in the backup process I will alter to make sure the dumped backup knows to drop any existing databases named "birds").

**Reply by Svardsten53** [reply]
That worked! Thanks for the support Patrick.

**Reply by mcguirepr89** [reply]
You are very welcome! I've adjusted the original post to show that the backup needs to drop the existing `birds` database
